### PR TITLE
feat: add operator-sdk build to kb layout

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description of the change:**

To maintain backwards compatibility and allow deprecation for it’s Kubebuilder Makefile equivalent "make docker-build", "operator-sdk build” command should work in Kubebuilder project layouts.

**Motivation for the change:**

Integration with KB

**Local Test**

```shell
 $ operator-sdk build test:0.0.1
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
INFO[0004] Building OCI image test:0.0.1                
Sending build context to Docker daemon  42.24MB
Step 1/14 : FROM golang:1.13 as builder
 ---> a80eb3fa7138
Step 2/14 : WORKDIR /workspace
 ---> Using cache
 ---> 82d53eb9405f
Step 3/14 : COPY go.mod go.mod
 ---> Using cache
 ---> 3e91f0628067
Step 4/14 : COPY go.sum go.sum
 ---> Using cache
 ---> 321ba688f9e9
Step 5/14 : RUN go mod download
 ---> Using cache
 ---> f8366218f3a0
Step 6/14 : COPY main.go main.go
 ---> Using cache
 ---> d14f4643b382
Step 7/14 : COPY api/ api/
 ---> Using cache
 ---> 5795655c887f
Step 8/14 : COPY controllers/ controllers/
 ---> Using cache
 ---> 39feeef283a9
Step 9/14 : RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
 ---> Using cache
 ---> c1578ecad410
Step 10/14 : FROM gcr.io/distroless/static:nonroot
 ---> 52ed073f70ec
Step 11/14 : WORKDIR /
 ---> Using cache
 ---> 41ad60687b26
Step 12/14 : COPY --from=builder /workspace/manager .
 ---> Using cache
 ---> f22dedf60d46
Step 13/14 : USER nonroot:nonroot
 ---> Using cache
 ---> 857ace31463a
Step 14/14 : ENTRYPOINT ["/manager"]
 ---> Using cache
 ---> 3563ba073474
Successfully built 3563ba073474
Successfully tagged test:0.0.1
INFO[0006] Operator build complete.                     
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-build $ 
```

### CODE REVIEW

File: cmd/operator-sdk/main.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func main() {
 	// a Kubebuilder project as a way to opt into the new KB CLI.
 	// TODO: Make the new KB CLI the default, once the integration is complete
 	// and deprecate "operator-sdk new" from the old CLI.
-	if kbutil.IsConfigExist() {
+	if kbutil.HasProjectFile() {
```
Comment (camilamacedo86) [inline cmd/operator-sdk/main.go:36]: Keep compatible with the name that will be used soon in kubebuilder for it.

File: cmd/operator-sdk/build/cmd.go [hunk 2]
```diff
@@ -94,18 +95,38 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
+
+	// If the project has the new kubebuilder layout
+	if kbutil.HasProjectFile() {
+		//makeDockerBuild := "make docker-build IMG="+image
+		cmd := exec.Command("make", "docker-build", "IMG="+image)
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			return err
+		}
+		fmt.Printf("%s\n", out)
```
Comment (joelanford) [inline cmd/operator-sdk/build/cmd.go:108]: Use `projutil.ExecCmd()` here so that output streams instead of buffers.

Comment (joelanford) [reply cmd/operator-sdk/build/cmd.go:108]: Actually, didn't we say that we want to have this command basically do the same thing as the legacy build except that:
1. the Dockerfile is in a different location
2. the Dockerfile is a multistage build, so there's no need to build the binary in advance.

File: internal/util/kubebuilder/project.go [hunk 3]
```diff
@@ -20,12 +20,14 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-const ConfigFile = "PROJECT"
+const (
+	configFile = "PROJECT"
```
Comment (camilamacedo86) [inline internal/util/kubebuilder/project.go:24]: Used just here so should not be global.

File: cmd/operator-sdk/build/cmd.go [hunk 4]
```diff
@@ -59,8 +61,15 @@ For example:
 		"Extra image build arguments as one string such as \"--build-arg https_proxy=$https_proxy\"")
 	buildCmd.Flags().StringVar(&imageBuilder, "image-builder", "docker",
 		"Tool to build OCI images. One of: [docker, podman, buildah]")
+
+	// Deprecated
 	buildCmd.Flags().StringVar(&goBuildArgs, "go-build-args", "",
```
Comment (joelanford) [inline cmd/operator-sdk/build/cmd.go:66]: Don't even add the flag if we're in the new project layout.
```suggestion
if !projutil.IsNewOperatorLayout() {
	buildCmd.Flags().StringVar(&goBuildArgs, "go-build-args", "", ...)
}
```

But don't deprecate it for the old project layouts. Once all the KB integration is in place, we'll deprecate the entirety of the old project layout.

File: cmd/operator-sdk/build/cmd.go [hunk 5]
```diff
@@ -126,20 +152,19 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
 		}
 	}
+	doImageBuild("build/Dockerfile", image)
+	return nil
+}
 
-	image := args[0]
-
+// doImageBuild will execute the build command for the Dockerfile and image informed
+func doImageBuild(dockerFilePath, image string) {
```
Comment (joelanford) [inline cmd/operator-sdk/build/cmd.go:160]: This should probably return an error, and we can bubble it up to `buildFunc`, where we can call `log.Fatal`

File: cmd/operator-sdk/main.go (hunk 1 above)
Comment (joelanford) [reply cmd/operator-sdk/main.go:36]: This seems out of scope for this PR. Can we make a separate PR so we can discuss each change on its merits separately?

File: internal/util/kubebuilder/project.go (hunk 3 above)
Comment (joelanford) [reply internal/util/kubebuilder/project.go:24]: Seems reasonable, but nit on the terminology. This is still global. But now it is unexported.

Ditto re: changes in this file should be done in a separate PR.

File: internal/util/projutil/project_util.go [hunk 6]
```diff
@@ -208,6 +216,12 @@ func IsOperatorGo() bool {
 	return err == nil || os.IsExist(err)
 }
 
+// todo(camilamacedo86): it should be removed when we no longer support the old scaffold layout
+// IsNewOperatorLayout return true when has the project file
+func IsNewOperatorLayout() bool {
+	return kbutil.HasProjectFile()
+}
```
Comment (joelanford) [inline internal/util/projutil/project_util.go:223]: This seems duplicative and unnecessary. Just use `kbutil.HasProjectFile()` wherever this would be used.

File: internal/util/projutil/project_util.go [hunk 7]
```diff
@@ -85,6 +87,12 @@ func MustInProjectRoot() {
 // TODO(hasbro17): Change this to check for go.mod
 // "build/Dockerfile" may not be present in all projects
 func CheckProjectRoot() error {
+	// If is the kubebuilder base layout
+	if kbutil.HasProjectFile() {
+		return nil
+	}
```
Comment (joelanford) [inline internal/util/projutil/project_util.go:93]: Is this change relevant to this PR? According to @hasbro17's TODO comment, we can have this check for `go.mod`, which would work in both cases.

File: cmd/operator-sdk/build/cmd.go [hunk 8]
```diff
@@ -95,14 +104,31 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
 
+	image := args[0]
 	projutil.MustInProjectRoot()
-	goBuildEnv := append(os.Environ(), "GOOS=linux")
 
+	if projutil.IsNewOperatorLayout() {
+		doImageBuild("Dockerfile", image)
+	} else {
+		if err := doLegacyBuild(image); err != nil {
+			return err
```
Comment (joelanford) [inline cmd/operator-sdk/build/cmd.go]: `log.Fatal` instead of returning an error, since this is not a usage error.

File: cmd/operator-sdk/build/cmd.go [hunk 9]
```diff
@@ -95,14 +104,31 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
+
+	return nil
+}
+
+// todo: remove when the legacy layout is no longer supported
+// Deprecated: Used just for the legacy layout
+// --
+// doLegacyBuild will build projects with the legacy layout.
+func doLegacyBuild(image string) error {
+	projutil.MustInProjectRoot()
```
Comment (joelanford) [inline cmd/operator-sdk/build/cmd.go:125]: We don't need to check for this again since we've already done it in line 98.

File: internal/util/projutil/project_util.go (hunk 7 above)
Comment (camilamacedo86) [reply internal/util/projutil/project_util.go:93]: I do not think that we should check for the go.mod. We can use the Project file to know if the project is Ansible, GO and/or Helm as we spoke before and remove all these checks.

Comment (joelanford) [reply internal/util/projutil/project_util.go:93]: Good point! And just to be clear, this change is relevant because we use `MustInProjectRoot` and before this change, that would fail in KB projects since it is looking for `./build/Dockefile`?

File: cmd/operator-sdk/build/cmd.go [hunk 10]
```diff
@@ -126,20 +154,19 @@ func buildFunc(cmd *cobra.Command, args []string) error {
...
-
 	if err := projutil.ExecCmd(buildCmd); err != nil {
-		log.Fatalf("Failed to output build image %s: %v", image, err)
+		return err
 	}
-
 	log.Info("Operator build complete.")
 	return nil
-}
+}
```
Comment (hasbro17) [inline cmd/operator-sdk/build/cmd.go:172]: Nit: leave newline

File: cmd/operator-sdk/build/cmd.go [hunk 11]
```diff
@@ -126,20 +154,19 @@ func buildFunc(cmd *cobra.Command, args []string) error {
 			log.Fatalf("Failed to build operator binary: %v", err)
 		}
 	}
+	return doImageBuild("build/Dockerfile", image)
+}
 
-	image := args[0]
-
+// doImageBuild will execute the build command for the Dockerfile and image informed
```
Comment (hasbro17) [inline cmd/operator-sdk/build/cmd.go:160]: Nit: Maybe rephrase. Not sure what image informed means.
```suggestion
// doImageBuild will execute the build command for the given Dockerfile path and image
```
